modules/network/src/main/java/org/apache/ignite/network/internal/netty/ConnectionManager.java

```diff
@@ -0,0 +1,167 @@
+            throw new IgniteInternalException("Failed to start server: " + cause.getMessage(), cause);
+        }
+        catch (InterruptedException e) {
+            throw new IgniteInternalException(
+                "Got interrupted while waiting for server to start: " + e.getMessage(),
+                e
+            );
+        }
+    }
+
+    /**
+     * @return Server local address.
+     */
+    public InetSocketAddress getLocalAddress() {
+        return server.address();
+    }
+
+    /**
+     * Get a {@link NettySender}, that sends data from this node to another node with the specified address.
+     * @param address Another node's address.
+     * @return Sender.
+     */
+    public CompletableFuture<NettySender> channel(InetSocketAddress address) {
+        NettySender channel = channels.get(address);
+
+        if (channel == null) {
+            NettyClient client = clients.computeIfAbsent(address, this::connect);
+            return client.sender();
+        }
+
+        return CompletableFuture.completedFuture(channel);
+    }
+
+    /**
+     * Callback that is called upon receiving of a new message.
+     *
+     * @param from Source of the message.
+     * @param message New message.
+     */
+    private void onMessage(InetSocketAddress from, NetworkMessage message) {
+        listeners.forEach(consumer -> {
+            consumer.accept(from, message);
+        });
+    }
+
+    /**
+     * Callback that is called upon new client connected to the server.
+     *
+     * @param channel Channel from client to this {@link #server}.
+     */
+    private void onNewIncomingChannel(SocketChannel channel) {
+        InetSocketAddress remoteAddress = channel.remoteAddress();
+        // TODO: there might be outgoing connection already
+        channels.put(remoteAddress, new NettySender(channel, serializationRegistry));
+    }
+
+    /**
+     * Create new client from this node to specified address.
+     *
+     * @param address Target address.
+     * @return New netty client.
+     */
+    private NettyClient connect(InetSocketAddress address) {
+        NettyClient client = new NettyClient(address.getHostName(), address.getPort(), serializationRegistry, (src, message) -> {
+            this.onMessage(src, message);
+        });
+
+        client.start().whenComplete((sender, throwable) -> {
+            if (throwable != null)
+                clients.remove(address);
+            else
+                channels.put(address, sender);
+        });
+
+        return client;
+    }
+
+    /**
+     * Add incoming message listener.
+     *
+     * @param listener Message listener.
+     */
+    public void addListener(BiConsumer<InetSocketAddress, NetworkMessage> listener) {
+        listeners.add(listener);
+    }
+
+    /**
+     * Stop server and all clients.
+     */
+    public void stop() {
+        // TODO: maybe add some flag that prohibts opening new connections from this moment?
```

Comment:
I think that this problem will be solved when you will have a single Bootstrap and worker group. Then you won't be able to accept connections if you close them.